### DISCUSSION TITLE
Integrate https://haveibeenpwned.com

### DISCUSSION THREAD

**Discussion by nguyenkims** [simple-login/app#400] [Feature Requests]
Frequently check using https://haveibeenpwned.com if an alias has been part of a data breach. If yes, inform users.

Moved from https://trello.com/c/kHODEw9w/124-have-i-been-pwned-check

**Comment by steinrueck** [comment]
I think this makes sense even for people using a Password Manager like 1Password. 1Password for example only checks for the registered email-address of the account, not for all login items.
[1Password multiple email addresses](https://1password.community/discussion/92691/please-support-searching-for-multiple-compromised-email-addresses)

**Comment by TheLastProject** [comment]
I'm willing to implement this. As I see it, it needs the following:

1. A job runner that:
    1. Gets a list of all known breaches and puts it in some database
    2. Start looping over all aliases
2. For each alias, set a "last checked" field and a list of pwned breaches
3. A button in the UI that can be clicked and shows a list of all breaches if any or something like that

To deal with [the rate limiting of HIBP](https://haveibeenpwned.com/API/v3#RateLimiting)
1. When starting the loop list the expected completion time
2. Support using multiple API keys to speed up the process
3. A setting that decides how often to start checking (if a check is still ungoing, print a warning and don't start the new check)

Does this seem reasonable?

**Reply by L235** [reply]
One other option, at least for SimpleLogin-owned domains, is to use the HIBP domain search: https://haveibeenpwned.com/DomainSearch. This will enable receiving a CSV file with all of the breaches and updates when any future breaches include emails on SimpleLogin domains.

**Reply by nguyenkims** [reply]
Thanks a lot Sylvia for taking this ticket! 

For the first version, I think we can have a global setting for how often to run a check on an alias (something like once per week is enough IMO). For avoiding the rate limiting issue, maybe we can just put a sleep of more than 1500 milliseconds between each alias check? 

We can use this cron.py file for storing the check job https://github.com/simple-login/app/blob/15610f1efcf9a6431a3a9d0cae052c21e778d45b/cron.py#L85

> When starting the loop list the expected completion time

I'm not sure to fully understand this, do you mean the script is aborted if it's run for too long?

> Support using multiple API keys to speed up the process

I think for the first version we can support a single API.

> A button in the UI that can be clicked and shows a list of all breaches if any or something like that

As the data breach information can be quite verbose, I think we can create a separate page for displaying this information. This page can be opened by a link in the "More" section below an alias, what do you think?

**Comment by jasperweiss** [comment]
Hasn't this been implemented already? It's still in the to-do column

**Reply by nguyenkims** [reply]
You're right, this has been implemented a while ago and should be marked as done. I just added the tag "done" :)

**Comment by JoeBoll** [comment]
Just curious, does the system check aliases made with custom domains?

**Reply by nguyenkims** [reply]
Yes it checks all aliases.

**Comment by DNCD** [comment]
Is HIBP implemented on version [3.4.0], or has it not been released yet? I added "HIBP_API_KEYS=[REDACTED]" to the .env. For some reason, it's not checking. I know that one of my aliases has been part of a data breach, but it's not showing up.

**Reply by nguyenkims** [reply]
The HIBP integration isn't included in the 3.4.0 indeed.

**Reply by DNCD** [reply]
Thank you so much for the reply. Looking forward to the new Update. Thank you!

**Comment by hiccups1980** [comment]
This is already implemented.

![pwned](https://github.com/user-attachments/assets/1398d9db-d37c-43f1-a24c-7993e273c293)
